src/FAudio_gstreamer.c

```diff
@@ -0,0 +1,850 @@
+#define SIZE_FROM_DST(sample) \
+	((sample) * voice->src.format->nChannels * sizeof(float))
+#define SAMPLES_FROM_SRC(len) \
+	((len) / voice->src.format->nChannels / (voice->src.format->wBitsPerSample / 8))
+#define SAMPLES_FROM_DST(len) \
+	((len) / voice->src.format->nChannels / sizeof(float))
+#define SIZE_SRC_TO_DST(len) \
+	((len) * (sizeof(float) / (voice->src.format->wBitsPerSample / 8)))
+#define SIZE_DST_TO_SRC(len) \
+	((len) / (sizeof(float) / (voice->src.format->wBitsPerSample / 8)))
+
+static int FAudio_GSTREAMER_RestartDecoder(FAudioGSTREAMER *gstreamer)
+{
+	GstEvent *event;
+
+	event = gst_event_new_flush_start();
+	if (!event)
+	{
+		return 0;
+	}
+	gst_pad_push_event(gstreamer->srcpad, event);
+
+	event = gst_event_new_flush_stop(TRUE);
+	if (!event)
+	{
+		return 0;
+	}
+	gst_pad_push_event(gstreamer->srcpad, event);
+
+	event = gst_event_new_segment(&gstreamer->segment);
+	if (!event)
+	{
+		return 0;
+	}
+	gst_pad_push_event(gstreamer->srcpad, event);
+
+	gst_element_set_state(gstreamer->pipeline, GST_STATE_PLAYING);
+
+	gstreamer->curBlock = ~0u;
+	gstreamer->prevBlock = ~0u;
+
+	return 1;
+}
+
+static int FAudio_GSTREAMER_CheckForBusErrors(FAudioVoice *voice)
+{
+	GstBus *bus;
+	GstMessage *msg;
+	GError *err = NULL;
+	gchar *debug_info = NULL;
+
+	bus = gst_pipeline_get_bus(GST_PIPELINE(voice->src.gstreamer->pipeline));
+	if (!bus)
+	{
+		return 0;
+	}
+
+	msg = gst_bus_pop_filtered(bus, GST_MESSAGE_ERROR);
+	if (msg)
+	{
+		switch(GST_MESSAGE_TYPE(msg))
+		{
+		case GST_MESSAGE_ERROR:
+			gst_message_parse_error(msg, &err, &debug_info);
+			LOG_ERROR(
+				voice->audio,
+				"Got gstreamer bus error from %s: %s (%s)",
+				GST_OBJECT_NAME(msg->src),
+				err->message,
+				debug_info ? debug_info : "none"
+			)
+			g_clear_error(&err);
+			g_free(debug_info);
+			gst_message_unref(msg);
+			return 1;
+		default:
+			gst_message_unref(msg);
+			break;
+		}
+	}
+
+	gst_object_unref(bus);
+
+	return 0;
+}
+
+static size_t FAudio_GSTREAMER_FillConvertCache(
+	FAudioVoice *voice,
+	FAudioBuffer *buffer,
+	size_t maxBytes
+) {
+	GstBuffer *src, *dst;
+	GstSample *sample;
+	GstMapInfo info;
+	size_t pulled, clipStartBytes, clipEndBytes, toCopyBytes;
+	GstAudioClippingMeta *cmeta;
+
+	LOG_FUNC_ENTER(voice->audio)
+
+	/* Write current block to input buffer, push to pipeline */
+	src = gst_buffer_new_allocate(
+		NULL,
+		voice->src.format->nBlockAlign,
+		NULL
+	);
+
+	if (	gst_buffer_fill(
+			src,
+			0,
+			buffer->pAudioData + (
+				voice->src.format->nBlockAlign *
+				voice->src.gstreamer->curBlock
+			),
+			voice->src.format->nBlockAlign
+		) != voice->src.format->nBlockAlign	)
+	{
+		LOG_ERROR(
+			voice->audio,
+			"for voice %p, failed to copy whole chunk into buffer",
+			voice
+		);
+		gst_buffer_unref(src);
+		return (size_t) -1;
+	}
+
+	if (	gst_pad_push(
+			voice->src.gstreamer->srcpad,
+			src
+		) != GST_FLOW_OK	)
+	{
+		LOG_ERROR(
+			voice->audio,
+			"for voice %p, failed to copy whole chunk into buffer",
```

Comment:
Copy and paste error from above and you probably want to also include the return value of `gst_pad_push()` in the message